scripts/install-build-tools.sh

```diff
@@ -6,19 +6,74 @@ green="\033[0;32m"
 cyan="\033[0;36m"
 end="\033[0m"
 
-set -e
+# for debugging, 'set -x' can be added to trace the commands
+set -ex
 
 SUDO=''
 if (( $EUID != 0 )); then
     echo -e "non-root user, sudo required"
     SUDO='sudo'
 fi
 
+# as of June 2024, python3.13 does not support pylint so limiting most recent version 3.12
+PYTHON_VERSIONS=("3.12" "3.11" "3.10")
+
+# Function to install Python and necessary packages
+install_python() {
+  local PY_VERS=$1
+  $SUDO apt install -y python${PY_VERS} python${PY_VERS}-distutils
+  wget https://bootstrap.pypa.io/get-pip.py
+  python${PY_VERS} get-pip.py
+  rm get-pip.py
+  python${PY_VERS} -m pip install pylint matplotlib
+}
+
+echo "Python3 at start of script version: $(python3 --version)"
+echo "OS Type: $OSTYPE"
+DEFAULT_PYTHON=`python3 --version | grep -E 'Python 3.[0-9]+'`
+
+
 if [[ "$OSTYPE" == "darwin"* ]]; then
   CPUS=$(sysctl -n hw.ncpu)
   # ensure development environment is set correctly for clang
   $SUDO xcode-select -switch /Library/Developer/CommandLineTools
-  brew install llvm@14 googletest google-benchmark lcov make wget cmake curl
+
+  echo "Running sudo with homebrew as root through a shell script is not allowed."
+  echo "These installs can be done manually in a terminal, however."
+
+  if ! command -v brew &>/dev/null; then
+    echo -e "${cyan}Homebrew is required to install dependencies.${end}"
+    exit 1
+  fi
+
+  brew install llvm@14 googletest google-benchmark lcov make wget cmake bash bc
+  brew upgrade bash
+
+  # Install one of the python versions if not already installed
+  if ! (echo "${DEFAULT_PYTHON}" | grep -E 'Python 3.1[0-2]'); then
+    FOUND=false
+    for PY_VERS in "${PY_VERSIONS[@]}"; do
+      if brew search python@${PY_VERS} | grep  "python@${PY_VERS}"; then
+        echo "Python$PY_VERS is available, so installing it"
+        brew install python@${PY_VERSIONS[0]} pylint python-matplotlib
+        export PATH="$(brew --prefix python@${PY_VERS})/bin:$PATH"
+        FOUND=true
+        break
+      fi
+    done
+
+    VERSION_SHORT=$(echo ${DEFAULT_PYTHON} | grep -E -o '3.[0-9]{1,2}')
+    echo "Python3 version: $VERSION_SHORT"
+    if [ "$FOUND" == false ]; then
+      echo "Python3 install with brew failed, attempted on these versions: ${PY_VERSIONS[@]}"
+      exit 1
+    elif [ echo "${VERSION_SHORT}" >= 3.13 ]; then
```

Comment:
Neither `test(1)` or `[[` support a `>=` operator portably. Unfortunately, because bash doesn't support non-integers, this cannot be easily replaced even with arithmetic evaluation. It could practically be done as the following:

```bash
(( (${VERSION_SHORT%.*} == 3) * (${VERSION_SHORT#*.} >= 13) ))
```

However, that's pretty magical. A common and clearer alternative is to pass things to `bc` which is specified by POSIX (and is what you used in `./scripts/pylint.sh`:

```bash
elif (( $(echo "${VERSION_SHORT} >= 3.13" | bc) )); then
```

Reply:
ACK - will use the second suggestion. Should it be brackets @HalosGhost not parentheses at the outermost part? as shown:
```console
elif [[ $(echo "${VERSION_SHORT} >= 3.13" | bc) ]]; then
```

Reply:
No, treating integers as truthy (as in C) is a special feature of arithmetic evaluation (`((`/`))`).

Reply:
ACK - fixed now thanks